File: test/language/union_types/java/union_types/UnionCompatibilityCheckTest.java
package union_types;

import static org.junit.jupiter.api.Assertions.*;
import org.junit.jupiter.api.Test;

import java.io.IOException;
import java.io.File;

import union_types.union_compatibility_check.UnionCompatibilityCheckVersion1;
import union_types.union_compatibility_check.UnionCompatibilityCheckVersion2;
import union_types.union_compatibility_check.UnionVersion1;
import union_types.union_compatibility_check.UnionVersion2;
import union_types.union_compatibility_check.CoordXY;
import union_types.union_compatibility_check.CoordXYZ;

import zserio.runtime.io.ByteArrayBitStreamReader;
import zserio.runtime.io.ByteArrayBitStreamWriter;
import zserio.runtime.io.FileBitStreamReader;
import zserio.runtime.io.FileBitStreamWriter;
import zserio.runtime.io.Writer;

public class UnionCompatibilityCheckTest
{
    @Test
    public void writeVersion1ReadVersion1() throws IOException
    {
        final UnionCompatibilityCheckVersion1 unionCompatibilityCheckVersion1 = create(createArrayVersion1());
        final UnionCompatibilityCheckVersion1 readUnionCompatibilityCheckVersion1 = writeReadVersion1(
                unionCompatibilityCheckVersion1);
        assertEquals(unionCompatibilityCheckVersion1, readUnionCompatibilityCheckVersion1);
    }

    @Test
    public void writeVersion1ReadVersion2() throws IOException
    {
        final UnionCompatibilityCheckVersion1 unionCompatibilityCheckVersion1 = create(createArrayVersion1());
        final UnionCompatibilityCheckVersion2 readUnionCompatibilityCheckVersion2 = writeReadVersion2(
                unionCompatibilityCheckVersion1);

        final UnionVersion2[] expectedArrayVersion2 = createArrayVersion2WithVersion1Fields();
        assertArrayEquals(expectedArrayVersion2, readUnionCompatibilityCheckVersion2.getArray());
        assertArrayEquals(expectedArrayVersion2, readUnionCompatibilityCheckVersion2.getPackedArray());
    }

    @Test
    public void writeVersion2ReadVersion1() throws IOException
    {
        final UnionCompatibilityCheckVersion2 unionCompatibilityCheckVersion2 =
                create(createArrayVersion2WithVersion1Fields());
        final UnionCompatibilityCheckVersion1 readUnionCompatibilityCheckVersion1 = writeReadVersion1(
                unionCompatibilityCheckVersion2);

        final UnionVersion1[] expectedArrayVersion1 = createArrayVersion1();
        assertArrayEquals(expectedArrayVersion1, readUnionCompatibilityCheckVersion1.getArray());
        assertArrayEquals(expectedArrayVersion1, readUnionCompatibilityCheckVersion1.getPackedArray());
    }

    @Test
    public void writeVersion2ReadVersion2() throws IOException
    {
        final UnionCompatibilityCheckVersion2 unionCompatibilityCheckVersion2 = create(createArrayVersion2());
        final UnionCompatibilityCheckVersion2 readUnionCompatibilityCheckVersion2 = writeReadVersion2(
                unionCompatibilityCheckVersion2);
        assertEquals(unionCompatibilityCheckVersion2, readUnionCompatibilityCheckVersion2);
    }

    @Test
    public void writeVersion1ReadVersion1File() throws IOException
    {
        final UnionCompatibilityCheckVersion1 unionCompatibilityCheckVersion1 = create(createArrayVersion1());
        final UnionCompatibilityCheckVersion1 readUnionCompatibilityCheckVersion1 = writeReadVersion1File(
                unionCompatibilityCheckVersion1, "version1_version1");
        assertEquals(unionCompatibilityCheckVersion1, readUnionCompatibilityCheckVersion1);
    }

    @Test
    public void writeVersion1ReadVersion2File() throws IOException
    {
        final UnionCompatibilityCheckVersion1 unionCompatibilityCheckVersion1 = create(createArrayVersion1());
        final UnionCompatibilityCheckVersion2 readUnionCompatibilityCheckVersion2 = writeReadVersion2File(
                unionCompatibilityCheckVersion1, "version1_version2");

        final UnionVersion2[] expectedArrayVersion2 = createArrayVersion2WithVersion1Fields();
        assertArrayEquals(expectedArrayVersion2, readUnionCompatibilityCheckVersion2.getArray());
        assertArrayEquals(expectedArrayVersion2, readUnionCompatibilityCheckVersion2.getPackedArray());
    }

    @Test
    public void writeVersion2ReadVersion1File() throws IOException
    {
        final UnionCompatibilityCheckVersion2 unionCompatibilityCheckVersion2 =
                create(createArrayVersion2WithVersion1Fields());
        final UnionCompatibilityCheckVersion1 readUnionCompatibilityCheckVersion1 = writeReadVersion1File(
                unionCompatibilityCheckVersion2, "version2_version1");

        final UnionVersion1[] expectedArrayVersion1 = createArrayVersion1();
        assertArrayEquals(expectedArrayVersion1, readUnionCompatibilityCheckVersion1.getArray());
        assertArrayEquals(expectedArrayVersion1, readUnionCompatibilityCheckVersion1.getPackedArray());
    }

    @Test
    public void writeVersion2ReadVersion2File() throws IOException
    {
        final UnionCompatibilityCheckVersion2 unionCompatibilityCheckVersion2 = create(createArrayVersion2());
        final UnionCompatibilityCheckVersion2 readUnionCompatibilityCheckVersion2 = writeReadVersion2File(
                unionCompatibilityCheckVersion2, "version2_version2");
        assertEquals(unionCompatibilityCheckVersion2, readUnionCompatibilityCheckVersion2);
    }

    private UnionCompatibilityCheckVersion1 create(UnionVersion1 array[])
    {
        return new UnionCompatibilityCheckVersion1(array, array);
    }

    private UnionCompatibilityCheckVersion2 create(UnionVersion2 array[])
    {
        return new UnionCompatibilityCheckVersion2(array, array);
    }

    private UnionVersion1[] createArrayVersion1()
    {
        return new UnionVersion1[] {
            createUnionVersion1(0),
            createUnionVersion1(1),
            createUnionVersion1(2),
            createUnionVersion1(3)
        };
    }

    private UnionVersion2[] createArrayVersion2WithVersion1Fields()
    {
        return new UnionVersion2[] {
            createUnionVersion2(0),
            createUnionVersion2(1),
            createUnionVersion2(2),
            createUnionVersion2(3)
        };
    }

    private UnionVersion2[] createArrayVersion2()
    {
        final UnionVersion2[] arrayVersion2WithVersion1Fields = createArrayVersion2WithVersion1Fields();
        return new UnionVersion2[] {
            arrayVersion2WithVersion1Fields[0],
            arrayVersion2WithVersion1Fields[1],
            arrayVersion2WithVersion1Fields[2],
            arrayVersion2WithVersion1Fields[3],
            createUnionCoordXYZ(4),
            createUnionCoordXYZ(5),
            createUnionCoordXYZ(6)
        };
    }

    private UnionVersion1 createUnionVersion1(int index)
    {
        final UnionVersion1 union = new UnionVersion1();
        if (index % 2 == 0)
            union.setCoordXY(new CoordXY(10L * index, 20L * index));
        else
            union.setText("text" + index);

        return union;
    }

    private UnionVersion2 createUnionVersion2(int index)
    {
        final UnionVersion2 union = new UnionVersion2();
        if (index % 2 == 0)
            union.setCoordXY(new CoordXY(10L * index, 20L * index));
        else
            union.setText("text" + index);

        return union;
    }

    private UnionVersion2 createUnionCoordXYZ(int index)
    {
        final UnionVersion2 union = new UnionVersion2();
        union.setCoordXYZ(new CoordXYZ(10L * index, 20L * index, 1.1 * index));
        return union;
    }

    private UnionCompatibilityCheckVersion1 writeReadVersion1(
            Writer unionCompatibilityCheck) throws IOException
    {
        final ByteArrayBitStreamWriter writer = new ByteArrayBitStreamWriter();
        unionCompatibilityCheck.write(writer);
        writer.close();

        final ByteArrayBitStreamReader reader = new ByteArrayBitStreamReader(writer.toByteArray());
        final UnionCompatibilityCheckVersion1 readUnion = new UnionCompatibilityCheckVersion1(reader);
        reader.close();

        return readUnion;
    }

    private UnionCompatibilityCheckVersion2 writeReadVersion2(
            Writer unionCompatibilityCheck) throws IOException
    {
        final ByteArrayBitStreamWriter writer = new ByteArrayBitStreamWriter();
        unionCompatibilityCheck.write(writer);
        writer.close();

        final ByteArrayBitStreamReader reader = new ByteArrayBitStreamReader(writer.toByteArray());
        final UnionCompatibilityCheckVersion2 readUnion = new UnionCompatibilityCheckVersion2(reader);
        reader.close();

        return readUnion;
    }

    private UnionCompatibilityCheckVersion1 writeReadVersion1File(
            Writer unionCompatibilityCheck, String variant) throws IOException
    {
        final File file = new File(BLOB_NAME_BASE + variant + ".blob");

        final FileBitStreamWriter writer = new FileBitStreamWriter(file);
        unionCompatibilityCheck.write(writer);
        writer.close();

        final FileBitStreamReader reader = new FileBitStreamReader(file);
        final UnionCompatibilityCheckVersion1 readUnion = new UnionCompatibilityCheckVersion1(reader);
        reader.close();

        return readUnion;
    }

    private UnionCompatibilityCheckVersion2 writeReadVersion2File(
            Writer unionCompatibilityCheck, String variant) throws IOException
    {
        final File file = new File(BLOB_NAME_BASE + variant + ".blob");

        final FileBitStreamWriter writer = new FileBitStreamWriter(file);
        unionCompatibilityCheck.write(writer);
        writer.close();

        final FileBitStreamReader reader = new FileBitStreamReader(file);
        final UnionCompatibilityCheckVersion2 readUnion = new UnionCompatibilityCheckVersion2(reader);
        reader.close();

        return readUnion;
    }

    private static final String BLOB_NAME_BASE = "union_compatibility_check_";
}

File: test/extensions/language/union_types/cpp/UnionCompatibilityCheckTest.cpp
#include "gtest/gtest.h"
#include "union_types/union_compatibility_check/UnionCompatibilityCheckVersion1.h"
#include "union_types/union_compatibility_check/UnionCompatibilityCheckVersion2.h"
#include "zserio/SerializeUtil.h"

namespace union_types
{
namespace union_compatibility_check
{

using allocator_type = UnionCompatibilityCheckVersion1::allocator_type;
template <typename T>
using vector_type = zserio::vector<T, allocator_type>;

class UnionCompatibilityCheckTest : public ::testing::Test
{
protected:
    template <typename T, typename ARRAY>
    void fill(T& unionCompatibilityCheck, const ARRAY& array)
    {
        unionCompatibilityCheck.setArray(array);
        unionCompatibilityCheck.setPackedArray(array);
    }

    template <typename T_READ, typename T_WRITE>
    T_READ writeRead(T_WRITE& unionCompatibilityCheck)
    {
        zserio::BitStreamWriter writer(bitBuffer);
        unionCompatibilityCheck.write(writer);

        zserio::BitStreamReader reader(writer.getWriteBuffer(), writer.getBitPosition(), zserio::BitsTag());
        return T_READ(reader);
    }

    template <typename T_READ, typename T_WRITE>
    T_READ writeReadFile(T_WRITE& unionCompatibilityCheck, const char* variant)
    {
        const std::string fileName = BLOB_NAME_BASE + variant + ".blob";
        zserio::serializeToFile(unionCompatibilityCheck, fileName);

        return zserio::deserializeFromFile<T_READ>(fileName);
    }

    template <typename UNION>
    UNION createUnion(uint32_t index)
    {
        UNION unionType;
        if (index % 2 == 0)
        {
            unionType.setCoordXY(CoordXY{10 * index, 20 * index});
        }
        else
        {
            unionType.setText("text" + zserio::toString<allocator_type>(index));
        }

        return unionType;
    }

    template <typename UNION>
    vector_type<UNION> createArrayVersion1()
    {
        return vector_type<UNION>{
                createUnion<UNION>(0), createUnion<UNION>(1), createUnion<UNION>(2), createUnion<UNION>(3)};
    }

    UnionVersion2 createUnionXYZ(uint32_t index)
    {
        UnionVersion2 unionType;
        unionType.setCoordXYZ(CoordXYZ{10 * index, 20 * index, 1.1 * index});
        return unionType;
    }

    vector_type<UnionVersion2> createArrayVersion2()
    {
        auto array = createArrayVersion1<UnionVersion2>();
        array.push_back(createUnionXYZ(4));
        array.push_back(createUnionXYZ(5));
        array.push_back(createUnionXYZ(6));
        return array;
    }

private:
    zserio::BitBuffer bitBuffer = zserio::BitBuffer(1024 * 8);

    static const std::string BLOB_NAME_BASE;
};

const std::string UnionCompatibilityCheckTest::BLOB_NAME_BASE =
        "language/union_types/union_compatibility_check_";

TEST_F(UnionCompatibilityCheckTest, writeVersion1ReadVersion1)
{
    UnionCompatibilityCheckVersion1 unionCompatibilityCheckVersion1;
    fill(unionCompatibilityCheckVersion1, createArrayVersion1<UnionVersion1>());

    auto readUnionCompatibilityCheckVersion1 =
            writeRead<UnionCompatibilityCheckVersion1>(unionCompatibilityCheckVersion1);
    ASSERT_EQ(unionCompatibilityCheckVersion1, readUnionCompatibilityCheckVersion1);
}

TEST_F(UnionCompatibilityCheckTest, writeVersion1ReadVersion1File)
{
    UnionCompatibilityCheckVersion1 unionCompatibilityCheckVersion1;
    fill(unionCompatibilityCheckVersion1, createArrayVersion1<UnionVersion1>());

    auto readUnionCompatibilityCheckVersion1 = writeReadFile<UnionCompatibilityCheckVersion1>(
            unionCompatibilityCheckVersion1, "version1_version1");
    ASSERT_EQ(unionCompatibilityCheckVersion1, readUnionCompatibilityCheckVersion1);
}

TEST_F(UnionCompatibilityCheckTest, writeVersion1ReadVersion2)
{
    UnionCompatibilityCheckVersion1 unionCompatibilityCheckVersion1;
    fill(unionCompatibilityCheckVersion1, createArrayVersion1<UnionVersion1>());

    auto readUnionCompatibilityCheckVersion2 =
            writeRead<UnionCompatibilityCheckVersion2>(unionCompatibilityCheckVersion1);
    const auto expectedArrayVersion2 = createArrayVersion1<UnionVersion2>();
    ASSERT_EQ(expectedArrayVersion2, readUnionCompatibilityCheckVersion2.getArray());
    ASSERT_EQ(expectedArrayVersion2, readUnionCompatibilityCheckVersion2.getPackedArray());
}

TEST_F(UnionCompatibilityCheckTest, writeVersion1ReadVersion2File)
{
    UnionCompatibilityCheckVersion1 unionCompatibilityCheckVersion1;
    fill(unionCompatibilityCheckVersion1, createArrayVersion1<UnionVersion1>());

    auto readUnionCompatibilityCheckVersion2 = writeReadFile<UnionCompatibilityCheckVersion2>(
            unionCompatibilityCheckVersion1, "version1_version2");
    const auto expectedArrayVersion2 = createArrayVersion1<UnionVersion2>();
    ASSERT_EQ(expectedArrayVersion2, readUnionCompatibilityCheckVersion2.getArray());
    ASSERT_EQ(expectedArrayVersion2, readUnionCompatibilityCheckVersion2.getPackedArray());
}

TEST_F(UnionCompatibilityCheckTest, writeVersion2ReadVersion1)
{
    UnionCompatibilityCheckVersion2 unionCompatibilityCheckVersion2;
    fill(unionCompatibilityCheckVersion2, createArrayVersion1<UnionVersion2>());

    auto readUnionCompatibilityCheckVersion1 =
            writeRead<UnionCompatibilityCheckVersion1>(unionCompatibilityCheckVersion2);
    const auto expectedArrayVersion1 = createArrayVersion1<UnionVersion1>();
    ASSERT_EQ(expectedArrayVersion1, readUnionCompatibilityCheckVersion1.getArray());
    ASSERT_EQ(expectedArrayVersion1, readUnionCompatibilityCheckVersion1.getPackedArray());
}

TEST_F(UnionCompatibilityCheckTest, writeVersion2ReadVersion1File)
{
    UnionCompatibilityCheckVersion2 unionCompatibilityCheckVersion2;
    fill(unionCompatibilityCheckVersion2, createArrayVersion1<UnionVersion2>());

    auto readUnionCompatibilityCheckVersion1 = writeReadFile<UnionCompatibilityCheckVersion1>(
            unionCompatibilityCheckVersion2, "version2_version1");
    const auto expectedArrayVersion1 = createArrayVersion1<UnionVersion1>();
    ASSERT_EQ(expectedArrayVersion1, readUnionCompatibilityCheckVersion1.getArray());
    ASSERT_EQ(expectedArrayVersion1, readUnionCompatibilityCheckVersion1.getPackedArray());
}

TEST_F(UnionCompatibilityCheckTest, writeVersion2ReadVersion2)
{
    UnionCompatibilityCheckVersion2 unionCompatibilityCheckVersion2;
    fill(unionCompatibilityCheckVersion2, createArrayVersion2());

    auto readUnionCompatibilityCheckVersion2 =
            writeRead<UnionCompatibilityCheckVersion2>(unionCompatibilityCheckVersion2);
    ASSERT_EQ(unionCompatibilityCheckVersion2, readUnionCompatibilityCheckVersion2);
}

TEST_F(UnionCompatibilityCheckTest, writeVersion2ReadVersion2File)
{
    UnionCompatibilityCheckVersion2 unionCompatibilityCheckVersion2;
    fill(unionCompatibilityCheckVersion2, createArrayVersion2());

    auto readUnionCompatibilityCheckVersion2 = writeReadFile<UnionCompatibilityCheckVersion2>(
            unionCompatibilityCheckVersion2, "version2_version2");
    ASSERT_EQ(unionCompatibilityCheckVersion2, readUnionCompatibilityCheckVersion2);
}

} // namespace union_compatibility_check
} // namespace union_types
